### ISSUE TITLE
Use context to obtain different outputs from similar inputs

### ISSUE THREAD

**Issue by msolefonte** [axa-group/nlp.js#553]
**Is your feature request related to a problem? Please describe.**
This is not related to any problem. I am quite sure it is a lacking feature but perhaps I have just missed it.

**Describe the solution you'd like**
I would like the NLP Manager to introduce contexts in a similar fashion as how they work in Dialogflow. Basically, in some situations, like in real life, you want to understand the same input differently based on the previous input. This can be, explained, for example, as an answer to a previous question. Let's see it with an example:

```
>>> I would like to create a new account. -> registerNewUser(), context = []
<<< Perfect. Can you give me your email? -> Answer to registerNewUser(). 
// {key: accountCreation, lifespan: 2} is pushed to the context. Lifespan is reduced with 
// each interaction till disappearing.
>>> myemail@mysite.com
// By default, a message containing only '%email%' should not result in any output, but, 
// if the context "accountCreation" is active, the result is different.
<<< Thank you for your answer. Proceeding.
```

I think Google has a neat example here: https://cloud.google.com/dialogflow/docs/contexts-overview

Context should also be used not only as a requirement but also as a recommendation (some priority in the classifying) or as a rejection (some results should not be shown under some contexts). I do not think there is any need to keep a context as is, with the lifespan, that could be handled by the user. However, the `manager.process()` function could accept an array of active context keys. 

```
manager.process('en', 'myemail@mysite.com', [accountCreation])
```

Of course, something similar should be introduced to the trainer. 

**Describe alternatives you've considered**
The clearest alternative to it is accept create an intent with the data you want to receive, checking the context once inside of the callback. For example, a message containing `%email%` should result in the invocation of the function `email()`. Then, inside of email, the context should be checked to consider the actions. I'll write an example in pseudocode.

```
email(email) {
  if accountCreation in context: accountCreationEmailReceived(email);
  elif login in context: loginEmailReceived(email);
  else fallbackFunction() // or nothing
}
```

The problem with this approach is that it is far harder to mantain, as it is less readable and natural. Also, sometimes, you want some rejections / affinities to work, which, if not happen, can result in something like this.

```
>>> Hello. My email is %email%. // No context
<<< hello()
>>> Hello. My email is %email% // Context = [accountCreation]
<<< accountCreationEmailReceived(email)
```

```
>>> Hello. // No context
<<< hello()
>>> Hello // Context = [alreadyGreeted]
<<< pass
```

These options are not possible to implement without a proper usage of context. You can still bypass it but it is, without any doubt, a poor patch.

**Additional context**
Again, I think that Dialogflow implements a perfect solution for the problem: https://cloud.google.com/dialogflow/docs/contexts-overview

**Comment by solyarisoftware** [comment]
Thanks Jesus for long reply! Above all thanks for your HUGE work with this project

I agree on the fact *NLP domains* and *DialogFlow contexts* are pretty different things

- *NLP domains*
  are a policy to split a multi-domain intents classifier in to small closed-domain classifier. That "divide et impera" is great immo and it differentiate NLP.js from competitors.

But, It seems to me that *domains* are removed in NLP.j v. 4. Why?
BTW, I proposed (privately, via mail) a CR to get a probability rank of the domain classifier. If makes sense I'd repropose here as CR.  

- *DialogFlow contexts*
  This is a way to manage a *states* of a conversation. I don't completely like the "Google" approach and I opensourced amy own small dialog manage, NaifJs, that's in a very alfa release (please forgive my bad js programming skills) and I just thought about an integration with NLp.js: https://github.com/solyarisoftware/naifjs/blob/master/doc/discussion.md#pattern-matching-agnostic

Thanks
giorgio

**Comment by shubhamtibra** [comment]
@jesus-seijas-sp . Thank you for the detailed explanation. I am using NLPjs and looking to add context support in chatbot.

Let's say the code suggested by you has been added in NLPjs.

Then the library will be able to filter some intents based on active context.
E.g. If the active context is "A", then don't consider an intent with input context "B".

I have added use case below. I would like to know if NLPjs can also help here.

Consider that I have added two intents in my bot:

1. Input context: "A", Training Phrase: "hi, how are you?"
2. Input context: NULL, Training Phrase: "hello, how are you doing?"

Now user says "hello, how are you doing?" and the active context is "A".
How do we choose which intent to trigger?

1st Intent has the same input context "A" but is not an exact match (context matching score is high, nlp matching score is low)
2nd Intent doesn't have input context but is an exact match of input (context matching score is low, nlp matching score is high)

What if I someone wanted to achieve this functionality where the code can give a balanced result consider both scores? Is it possible to achieve using NLPjs and Microsoft Framework? Or can this be added in NLPjs similar to what you have suggested?
